oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/registration/ClientRegistrations.java

```diff
@@ -100,6 +107,66 @@ public static ClientRegistration.Builder fromOidcIssuerLocation(String issuer) {
 		return getBuilder(issuer, oidc(URI.create(issuer)));
 	}
 
+	/**
+	 * Creates a {@link ClientRegistration.Builder} using the provided <a href=
+	 * "https://openid.net/specs/openid-connect-core-1_0.html#IssuerIdentifier">Issuer</a>
+	 * and {@link Function metadata resolver} to obtain the values used to initialize the
+	 * {@link ClientRegistration.Builder}.
+	 *
+	 * <p>
+	 * This method differs from {@link #fromOidcIssuerLocation(String)} in that the
+	 * <a href=
+	 * "https://openid.net/specs/openid-connect-discovery-1_0.html#ProviderConfigurationResponse">OpenID
+	 * Provider Configuration Response</a> is provided by the metadata resolver, which
+	 * allows flexibility in determining how the metadata is obtained. For example, you
+	 * can use any HTTP client configured with user-defined proxy settings, obtain
+	 * metadata from a file on the filesystem, etc.
+	 * </p>
+	 *
+	 * <p>
+	 * The default metadata resolver used by {@link #fromOidcIssuerLocation(String)} uses
+	 * a {@code RestTemplate} to make an <a href=
+	 * "https://openid.net/specs/openid-connect-discovery-1_0.html#ProviderConfigurationRequest">OpenID
+	 * Provider Configuration Request</a> to the metadata endpoint. For example, if the
+	 * issuer provided is "https://example.com", then an "OpenID Provider Configuration
+	 * Request" will be made to "https://example.com/.well-known/openid-configuration".
+	 * The result is expected to be an "OpenID Provider Configuration Response". The
+	 * default metadata resolver is as follows:
+	 * </p>
+	 * <pre>
+	 * RestTemplate rest = new RestTemplate();
+	 * ParameterizedTypeReference&lt;Map&lt;String, Object&gt;&gt; typeReference = new ParameterizedTypeReference&lt;Map&lt;String, Object&gt;&gt;() {
+	 * };
+	 * Function&lt;URI, Map&lt;String, Object&gt;&gt; metadataResolver = (metadataUri) -> {
+	 *     RequestEntity&lt;Void&gt; request = RequestEntity.get(metadataUri).build();
+	 *     return rest.exchange(request, typeReference).getBody();
+	 * };
+	 * </pre>
+	 *
+	 * <p>
+	 * Example usage:
+	 * </p>
+	 * <pre>
+	 * ClientRegistration registration =
+	 *     ClientRegistrations.fromOidcIssuerMetadata("https://example.com", metadataResolver)
+	 *         .clientId("client-id")
+	 *         .clientSecret("client-secret")
+	 *         .build();
+	 * </pre>
+	 * @param issuer the <a href=
+	 * "https://openid.net/specs/openid-connect-core-1_0.html#IssuerIdentifier">Issuer</a>
+	 * @param metadataResolver the {@code Function} used to resolve metadata in an
+	 * application-specific way
+	 * @return a {@link ClientRegistration.Builder} that was initialized by the OpenID
+	 * Provider Configuration.
+	 * @since 5.8
+	 */
+	public static ClientRegistration.Builder fromOidcIssuerMetadata(String issuer,
+			Function<URI, Map<String, Object>> metadataResolver) {
```

Comment:
Thanks for the feedback @jzheaux.

I see your point regarding...

> It also feels a little odd passing behavior as a parameter

I agree that in certain scenarios it doesn't make sense from a design standpoint, however, there are exceptions to this principle. `RestTemplate.execute()` accepts `RequestCallback` and `ResponseExtractor` and Reactive streams accept `Function`, `BiFunction`, etc. that all provide behavioural logic.

Regardless, your points have swayed me away from the overloaded method strategy and it feels more that the behaviour should be replaced as a whole within the utility component by exposing a setter. So I'm proposing we expose a `static` setter instead of the overloaded methods:

```java
private static final Function<URI, Map<String, Object>> DEFAULT_METADATA_RESOLVER = (metadataUri) -> {
	RequestEntity<Void> request = RequestEntity.get(metadataUri).build();
	return rest.exchange(request, typeReference).getBody();
};

private static Function<URI, Map<String, Object>> metadataResolver = DEFAULT_METADATA_RESOLVER;

public static void setMetadataResolver(Function<URI, Map<String, Object>> metadataResolver) {
	ClientRegistrations.metadataResolver = metadataResolver;
}
```

With this design, the consuming application would need to call `setMetadataResolver()` in a `static` initializing block in order to change the behaviour before `ClientRegistrations.fromIssuerLocation()` is called from another component, e.g. Spring Boot's `OAuth2ClientPropertiesRegistrationAdapter`.

I realize this may not be the "cleanest" design but it is by far the most simple compared to creating a new component or exposing a new builder. This simple design can address the Spring Boot issues that arise when the `issuer-uri` property is used and there are firewall/proxy related issues.

What are your thoughts on this approach?

Reply:
One issue that might come up with this approach is integrating with Spring Boot. In the past, we've run into issues when trying to configure `SecurityContextHolder#setContextHolderStrategy`. As of right now, there are no plans (I suppose) to have Spring Boot apply a `RestTemplate` instance to `ClientRegistrations#setMetadataResolver`; however, I think we'd see the same resistance to the idea should that come up.

Also related to Boot's concern about configuring `SecurityContextHolder`, introducing static configurability to `ClientRegistrations` means that it can't be configured differently for multiple application contexts.

> a `static` initializing block

My concern here is that it would deviate from the general guidance we've given so far about `RestTemplate` configuration being `@Bean`-based. In an initializing block, there would be no opportunity to pick up a Spring-managed `RestTemplate` instance and apply it.

If there is just one application context, the code can refer to `ClientRegistrations#setMetadataResolver` in an `@Autowired`-annotated method instead of being in a static initializing block.

> `setMetadataResolver`

I don't have a strong opinion here, but my preference is `setRestOperations` since that's what is exposed in most other places in Spring Security. I understand that there may be additional use cases this PR is trying to address, though.